### PR TITLE
tf.TensorLike

### PR DESCRIPTION
**Pull request by rahul-kamat**
Define `tf.TensorLike` to be a Union of all types which have a registered Tensor conversion function.

Use tf.TensorLike to annotate function arguments and treat them as if they were a tensor.

### Examples
Using `tf.TensorLike` to annotate function arguments:
```python
def foo(t: tf.TensorLike):
  return t

x = foo(1) # x = tf.Tensor(1, shape=(), dtype=int32)
```

Without `tf.TensorLike` annotation:
```python
def foo(t):
  return t

x = foo(1) # x = 1
```

Note: The input will not actually be converted to a Tensor, but using `tf.TensorLike` will allow input to be defined as a Tensor.

### CODE REVIEW

File: tensorflow/python/types/core.py [hunk 1]
```diff
@@ -58,3 +62,7 @@ class Value(Tensor):
 
   def numpy(self):
     pass
+
+
+TensorLike = Union[Tensor, int, float, bool, str, tuple, list, np.ndarray]
+tf_export("TensorLike").export_constant(__name__, "TensorLike")
```
Comment (mdanatg) [inline tensorflow/python/types/core.py:68]: Let's use "types.experimental.TensorLike" for now. We also need to update the goldens for api_test. Lastly, we'll need to document the symbol.

File: tensorflow/python/types/core.py [hunk 2]
```diff
@@ -58,3 +62,7 @@ class Value(Tensor):
 
   def numpy(self):
     pass
+
+
+TensorLike = Union[Tensor, int, float, bool, str, tuple, list, np.ndarray]
```
Comment (mdanatg) [inline tensorflow/python/types/core.py:67]: I think we also need to add the complex type.

**Review by mdanatg — COMMENTED** [review]
Looks good, just need a bit more boilerplate for the symbol export.

**Comment by edloper** [comment]
There are other types that are convertible to tensor -- should we include any/some/all of them?  E.g.:

* builtins.range
* Operation
* TensorShape
* Dimension
* IndexedSlices
* BaseResourceVariable, RefVariable
* DistributedVariable, MirroredVariable, Mirrored, SyncOnReadVariable
* AggregatingVariable
* Asset
* PackedVarAndDevice

**Comment by mdanatg** [comment]
I think that makes sense; we should consider each instance carefully. I think we could add builtins.range right away (better yet, as `Iterable`). For the others, we should have a closer look and check whether we should name them explicitly or via a supertype (e.g. Variable will be TensorLike via the TensorConvertible protocol).

File: tensorflow/python/types/core.py (hunk 1 above)
Comment (rahul-kamat) [reply tensorflow/python/types/core.py]: I used "experimental.TensorLike" instead to allow the goldens to be updated.

**Review by mdanatg — DISMISSED** [review]

**Comment by gbaned** [comment]
@rahul-kamat Can you please address Ubuntu Sanity errors? Thanks!

**Comment by rahul-kamat** [comment]
TensorLike includes Tensor, primitive variable, tuple, list, and ndarray types. We can add other types that are convertible to Tensor to this union.

**Review by mdanatg — APPROVED** [review]

**Comment by byronyi** [comment]
@rahul-kamat @mdanatg @mihaimaruseac This patch fails API compatibility tests for Python >= 3.7:

```
$ bazel run --repo_env=PYTHON_BIN_PATH=$(which python3) tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=181
INFO: Reading rc options for 'run' from /data00/home/byronyi/tensorflow/.bazelrc:
  Inherited 'common' options: --experimental_repo_remote_exec
INFO: Reading rc options for 'run' from /data00/home/byronyi/tensorflow/.bazelrc:
  Inherited 'build' options: --apple_platform_type=macos --define framework_shared_object=true --define open_source_build=true --java_toolchain=//third_party/toolchains/java:tf_java_toolchain --host_java_toolchain=//third_party/toolchains/java:tf_java_toolchain --define=use_fast_cpp_protos=true --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --noincompatible_remove_legacy_whole_archive --noincompatible_prohibit_aapt1 --enable_platform_specific_config --config=v2
INFO: Found applicable config definition build:v2 in file /data00/home/byronyi/tensorflow/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Found applicable config definition build:linux in file /data00/home/byronyi/tensorflow/.bazelrc: --copt=-w --define=PREFIX=/usr --define=LIBDIR=$(PREFIX)/lib --define=INCLUDEDIR=$(PREFIX)/include --cxxopt=-std=c++14 --host_cxxopt=-std=c++14 --config=dynamic_kernels
INFO: Found applicable config definition build:dynamic_kernels in file /data00/home/byronyi/tensorflow/.bazelrc: --define=dynamic_loaded_kernels=true --copt=-DAUTOLOAD_DYNAMIC_KERNELS
WARNING: /data00/home/byronyi/tensorflow/tensorflow/core/BUILD:1720:1: in linkstatic attribute of cc_library rule //tensorflow/core:lib_internal: setting 'linkstatic=1' is recommended if there are no object files. Since this rule was created by the macro 'cc_library', the error might have been caused by the macro implementation
WARNING: /data00/home/byronyi/tensorflow/tensorflow/core/BUILD:2132:1: in linkstatic attribute of cc_library rule //tensorflow/core:framework_internal: setting 'linkstatic=1' is recommended if there are no object files. Since this rule was created by the macro 'tf_cuda_library', the error might have been caused by the macro implementation
WARNING: /data00/home/byronyi/tensorflow/tensorflow/core/BUILD:1745:1: in linkstatic attribute of cc_library rule //tensorflow/core:lib_headers_for_pybind: setting 'linkstatic=1' is recommended if there are no object files. Since this rule was created by the macro 'cc_library', the error might have been caused by the macro implementation
WARNING: /data00/home/byronyi/tensorflow/tensorflow/python/BUILD:4666:1: in py_library rule //tensorflow/python:standard_ops: target '//tensorflow/python:standard_ops' depends on deprecated target '//tensorflow/python/ops/distributions:distributions': TensorFlow Distributions has migrated to TensorFlow Probability (https://github.com/tensorflow/probability). Deprecated copies remaining in tf.distributions will not receive new features, and will be removed by early 2019. You should update all usage of `tf.distributions` to `tfp.distributions`.
WARNING: /data00/home/byronyi/tensorflow/tensorflow/python/BUILD:115:1: in py_library rule //tensorflow/python:no_contrib: target '//tensorflow/python:no_contrib' depends on deprecated target '//tensorflow/python/ops/distributions:distributions': TensorFlow Distributions has migrated to TensorFlow Probability (https://github.com/tensorflow/probability). Deprecated copies remaining in tf.distributions will not receive new features, and will be removed by early 2019. You should update all usage of `tf.distributions` to `tfp.distributions`.
INFO: Analyzed target //tensorflow/tools/api/tests:api_compatibility_test (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //tensorflow/tools/api/tests:api_compatibility_test up-to-date:
  bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
INFO: Elapsed time: 0.707s, Critical Path: 0.02s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow/tools/api/tests:api_compatibility_test
-----------------------------------------------------------------------------
Running tests under Python 3.8.4: /usr/bin/python3
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibility
ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

E0726 00:08:50.171075 140334335260480 api_compatibility_test.py:280] TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

ERROR:tensorflow:1 differences found between API and golden.
E0726 00:08:50.171331 140334335260480 api_compatibility_test.py:281] 1 differences found between API and golden.
WARNING:tensorflow:Golden file update requested!
All test failures have been skipped, see the logs for detected diffs.
This test is now going to write new golden files.
Make sure to package the updates together with your change.

You will need an explicit API approval. This may take longer than a normal
review.

W0726 00:08:50.171641 140334335260480 api_compatibility_test.py:285] Golden file update requested!
All test failures have been skipped, see the logs for detected diffs.
This test is now going to write new golden files.
Make sure to package the updates together with your change.

You will need an explicit API approval. This may take longer than a normal
review.

INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibility): 2.37s
I0726 00:08:50.178608 140334335260480 test_util.py:1972] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibility): 2.37s
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibility
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

E0726 00:08:52.495300 140334335260480 api_compatibility_test.py:280] TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

ERROR:tensorflow:1 differences found between API and golden.
E0726 00:08:52.495720 140334335260480 api_compatibility_test.py:281] 1 differences found between API and golden.
WARNING:tensorflow:Golden file update requested!
All test failures have been skipped, see the logs for detected diffs.
This test is now going to write new golden files.
Make sure to package the updates together with your change.

You will need an explicit API approval. This may take longer than a normal
review.

W0726 00:08:52.495843 140334335260480 api_compatibility_test.py:285] Golden file update requested!
All test failures have been skipped, see the logs for detected diffs.
This test is now going to write new golden files.
Make sure to package the updates together with your change.

You will need an explicit API approval. This may take longer than a normal
review.

INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV1): 2.32s
I0726 00:08:52.503326 140334335260480 test_util.py:1972] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV1): 2.32s
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
INFO:tensorflow:No differences found between API and golden.
I0726 00:08:54.558901 140334335260480 api_compatibility_test.py:307] No differences found between API and golden.
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV2): 2.06s
I0726 00:08:54.572412 140334335260480 test_util.py:1972] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV2): 2.06s
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessage): 0.27s
I0726 00:08:54.845571 140334335260480 test_util.py:1972] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessage): 0.27s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV1): 0.3s
I0726 00:08:55.141913 140334335260480 test_util.py:1972] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV1): 0.3s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV2): 0.29s
I0726 00:08:55.432852 140334335260480 test_util.py:1972] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV2): 0.29s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
----------------------------------------------------------------------
Ran 7 tests in 7.624s

OK (skipped=1)
```

And the diff made:

```
$ git diff
diff --git a/tensorflow/tools/api/golden/v1/tensorflow.types.experimental.pbtxt b/tensorflow/tools/api/golden/v1/tensorflow.types.experimental.pbtxt
index d1b54df0c8..0ec48cd220 100644
--- a/tensorflow/tools/api/golden/v1/tensorflow.types.experimental.pbtxt
+++ b/tensorflow/tools/api/golden/v1/tensorflow.types.experimental.pbtxt
@@ -2,6 +2,6 @@ path: "tensorflow.types.experimental"
 tf_module {
   member {
     name: "TensorLike"
-    mtype: "typing.Union"
+    mtype: "<class \'typing._GenericAlias\'>"
   }
 }
diff --git a/tensorflow/tools/api/golden/v2/tensorflow.types.experimental.pbtxt b/tensorflow/tools/api/golden/v2/tensorflow.types.experimental.pbtxt
index d1b54df0c8..0ec48cd220 100644
--- a/tensorflow/tools/api/golden/v2/tensorflow.types.experimental.pbtxt
+++ b/tensorflow/tools/api/golden/v2/tensorflow.types.experimental.pbtxt
@@ -2,6 +2,6 @@ path: "tensorflow.types.experimental"
 tf_module {
   member {
     name: "TensorLike"
-    mtype: "typing.Union"
+    mtype: "<class \'typing._GenericAlias\'>"
   }
 }
```

**Comment by byronyi** [comment]
In the diff page of this PR, the `mtype` of `TensorLike` is `typing._GenericAlias`.

Yet it changes to `typing.Union` when this PR is merged in a366858377a2942b691c8664c10e142d8bca8452.

**Comment by mdanatg** [comment]
The API compatibility test was fixed recently to avoid this error. Are you running the test at head?

**Pull request merged**